commit 44b372d8a099a7042f9f17ebd4941050e38c1773
Author: Anton Blanchard <anton@samba.org>
Date:   Sun Jun 24 18:24:44 2012 +0000

    powerpc/vio: Separate vio bus probe and device probe
    
    Similar to PCI, separate the bus probe from device probe. This allows
    us to attach bus notifiers for DMA debug and IOMMU fault injection
    before devices have been probed.
    
    Signed-off-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>

diff --git a/arch/powerpc/kernel/vio.c b/arch/powerpc/kernel/vio.c
index 7d05fe2b5637..3bb5be5da31c 100644
--- a/arch/powerpc/kernel/vio.c
+++ b/arch/powerpc/kernel/vio.c
@@ -1497,12 +1497,18 @@ static int __init vio_bus_init(void)
 	if (firmware_has_feature(FW_FEATURE_CMO))
 		vio_cmo_bus_init();
 
+	return 0;
+}
+postcore_initcall(vio_bus_init);
+
+static int __init vio_device_init(void)
+{
 	vio_bus_scan_register_devices("vdevice");
 	vio_bus_scan_register_devices("ibm,platform-facilities");
 
 	return 0;
 }
-__initcall(vio_bus_init);
+device_initcall(vio_device_init);
 
 static ssize_t name_show(struct device *dev,
 		struct device_attribute *attr, char *buf)